src/service/troopMovements.ts

```diff
@@ -0,0 +1,128 @@
+            countryInfo.set(parseInt(country.id), country);
+            acc.push(buildUnitChange(country, prev));
+        }
+        return acc;
+    }, []);
+}
+
+export function detectOrigin(changedWorld: Units[]): Units[] {
+    const duplicatesToRemove: number[] = [];
+
+    const addedPrevCountry = changedWorld.map((c1) => {
+        const allies = changedWorld.find((c2) =>
+            linkMovement(c1.delta_allies, c2.delta_allies)
+        );
+        if (allies?.country && allies.country !== c1.country) {
+            c1.previous_country = allies?.country;
+            duplicatesToRemove.push(c1.previous_country);
+        }
+
+        const axis = changedWorld.find((c2) =>
+            linkMovement(c1.delta_axis, c2.delta_axis)
+        );
+        if (axis?.country && axis.country !== c1.country) {
+            c1.previous_country = axis?.country;
+            duplicatesToRemove.push(c1.previous_country);
+        }
+        return c1;
+    });
+
+    return addedPrevCountry.filter((country) => {
+        return !duplicatesToRemove.includes(country.country);
+    });
+}
+
+export async function prepareAndSendMessage(
+    _client: Client,
+    changedUnits: Units[]
+) {
+    const guildRepository = AppDataSource.getRepository(Guild);
+    const guilds = await guildRepository.find();
+    guilds.forEach(async (guild) => {
+        if (guild.troop_movement_channel) {
+            const channel: TextChannel = _client.channels.cache.get(
+                guild.troop_movement_channel
+            ) as TextChannel;
+
+            changedUnits.forEach(async (unit) => {
+                sendTMMessage(unit, channel);
+            });
+        }
+    });
+}
+
+function sendTMMessage(unit: Units, channel: TextChannel) {
+    const country = countryInfo.get(unit.country);
+    const icon = getIcon(country?.controlTeam ?? 0);
+    const control = getControl(country);
+    const team = unit.delta_allies !== 0 ? 1 : 2;
+
+    const embed = new MessageEmbed()
+        .setTitle(`${icon} ${country?.name} (${control}%) [${country?.region}]`)
+        .setDescription(
+            unit.previous_country
+                ? getEmbedDesc(unit)
+                : `Change in unit count detected:`
+        )
+        .addField(
+            "Allied Forces",
+            `${unit.allies} (${unit.delta_allies})`,
+            true
+        )
+        .addField("Axis Forces", `${unit.axis} (${unit.delta_axis})`, true)
+        .setColor(unit.previous_country ? getColor(team) : Color.BLUE);
+    channel?.send({ embeds: [embed] });
+}
+
+function getEmbedDesc(unit: Units) {
+    const prevCountry = countryInfo.get(unit.previous_country ?? 0);
+    const prevIcon = getIcon(prevCountry?.controlTeam ?? 0);
+    const teamName = unit.delta_allies !== 0 ? "Allied" : "Axis";
+    const control = getControl(prevCountry);
+
+    return `${teamName} units have arrived from ${prevIcon} ${prevCountry?.name} (${control}%) (?-?mins)`;
```

Comment:
the utility function to compute travel times are available in `map/geometry` and `map/util`